src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java

```diff
@@ -0,0 +1,203 @@
+		} else {
+			pToastComponent.blit(pPoseStack, 0, 17, 176, 0, 20, 20);
+			pToastComponent.getMinecraft().font.draw(pPoseStack, this.type.getDisplayName(), 30.0F, 7.0F, 0xff0000);
+			pToastComponent.getMinecraft().font.draw(pPoseStack, new TextComponent("-" + String.valueOf(this.difference)), 30.0F, 18.0F, 0x000000);
+		}
+		
+		posestack = RenderSystem.getModelViewStack();
+		posestack.pushPose();
+		
+		//scale for the grist-icon since it doesn't inherit the background's scale for some reason.
+		posestack.scale(0.8F, 0.8F, 1.0F);
+		
+		RenderSystem.applyModelViewMatrix();
+		
+		//draws the grist icon.
+		this.drawIcon(100, 4, type.getIcon());
+		
+		posestack.popPose();
+		RenderSystem.applyModelViewMatrix();
+		
+		return pTimeSinceLastVisible - this.lastChanged >= DISPLAY_TIME ? Toast.Visibility.HIDE : Toast.Visibility.SHOW;
+		
+		
+	}
+	
+	//modified version of drawIcon() from MinestuckScreen.java
+	private void drawIcon(int x, int y, ResourceLocation icon)
+	{
+		if(icon == null || Minecraft.getInstance() == null)
+			return;
+		
+		RenderSystem.setShader(GameRenderer::getPositionTexShader);
+		RenderSystem.setShaderTexture(0, icon);
+		
+		float scale = (float) 1 / 16;
+		
+		int iconX = 16;
+		int iconY = 16;
+		int iconU = 0;
+		int iconV = 0;
+		
+		BufferBuilder render = Tesselator.getInstance().getBuilder();
+		render.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX);
+		render.vertex(x, y + iconY, 0D).uv((iconU) * scale, (iconV + iconY) * scale).endVertex();
+		render.vertex(x + iconX, y + iconY, 0D).uv((iconU + iconX) * scale, (iconV + iconY) * scale).endVertex();
+		render.vertex(x + iconX, y, 0D).uv((iconU + iconX) * scale, (iconV) * scale).endVertex();
+		render.vertex(x, y, 0D).uv((iconU) * scale, (iconV) * scale).endVertex();
+		Tesselator.getInstance().end();
+	}
+	
+	//adds pDifference to the toast's current grist value.
+	private void addGrist(long pDifference) {
+		
+		this.difference += pDifference;
+		this.changed = true;
+		
+	}
+	
+	//Updates the grist value of any existing toasts, and if there aren't any of the same type, it instantiates a new one. NEVER use addToast() directly when adding a grist toast, ALWAYS use this method.
+	public static void addOrUpdate(ToastComponent pToastGui, GristType pType, long pDifference, EnumSource pSource, boolean pIncrease) {
+		
+		//try to find an existing toast with the same properties as the one being added.
+		GristToast gristToast = pToastGui.getToast(GristToast.class, pType.getTranslationKey() + pSource + String.valueOf(pIncrease));
+		
+		//if none can be found, add a new toast. If an existing toast *can* be found, update its grist amount using pDifference.
+		if (gristToast == null) {
+			pToastGui.addToast(new GristToast(pType, pDifference, pSource, pIncrease));
+		} else {
+			gristToast.addGrist(pDifference);
+		}
+		
+	}
+	
+	public static void sendGristMessage(Player player, GristSet set, GristToast.EnumSource source, boolean increase)
```

Comment:
You're not actually using the player argument for anything but the `isLocalPlayer()` check. You could probably remove the check entirely, but if you still want a check to that effect, you might be able to use `minecraft.isSameThread()`?

Reply:
It was mostly just a sanity check, so I've removed it.